src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs

```diff
@@ -0,0 +1,243 @@
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        public async Task<ResourceWrapper> GetAsync(ResourceKey key, CancellationToken cancellationToken)
+        {
+            using (var connection = new SqlConnection(_configuration.ConnectionString))
+            {
+                await connection.OpenAsync(cancellationToken);
+
+                using (SqlCommand command = connection.CreateCommand())
+                {
+                    command.Parameters.AddWithValue("@resourceTypeId", _model.GetResourceTypeId(key.ResourceType));
+                    command.Parameters.AddWithValue("@resourceId", key.Id);
+
+                    bool versionedRead = !string.IsNullOrEmpty(key.VersionId);
+
+                    if (versionedRead)
+                    {
+                        command.CommandText = @"
+                            SELECT Version, LastUpdated, IsDeleted, IsHistory, RawResource 
+                            FROM dbo.Resource
+                            WHERE ResourceTypeId = @resourceTypeId AND ResourceId = @resourceId AND Version = @version";
+
+                        if (!int.TryParse(key.VersionId, out var version))
+                        {
+                            return null;
+                        }
+
+                        command.Parameters.AddWithValue("@version", version);
+                    }
+                    else
+                    {
+                        command.CommandText = @"
+                            SELECT Version, LastUpdated, IsDeleted, IsHistory, RawResource 
+                            FROM dbo.Resource
+                            WHERE ResourceTypeId = @resourceTypeId AND ResourceId = @resourceId AND IsHistory = 0";
+                    }
+
+                    using (SqlDataReader sqlDataReader = await command.ExecuteReaderAsync(CommandBehavior.SequentialAccess, cancellationToken))
+                    {
+                        if (!sqlDataReader.Read())
+                        {
+                            return null;
+                        }
+
+                        int version = sqlDataReader.GetInt32(0);
+                        DateTime lastModified = sqlDataReader.GetDateTime(1);
+                        var isDeleted = sqlDataReader.GetBoolean(2);
+                        var isHistory = sqlDataReader.GetBoolean(3);
+
+                        string rawResource;
+
+                        using (var dataStream = sqlDataReader.GetStream(4))
+                        using (var gzipStream = new GZipStream(dataStream, CompressionMode.Decompress))
+                        using (var reader = new StreamReader(gzipStream, ResourceEncoding))
+                        {
+                            rawResource = await reader.ReadToEndAsync();
+                        }
+
+                        return new ResourceWrapper(
```

Comment:
Any reason we're not using the `ResourceWrapperFactory`?

Reply:
Because it's coupled with search parameter extraction. 
Separately, I think a we could come up with a better abstration for the contract, at least for the response. It was convenient for CosmosDB because it's the deserialization of the document, but doesn't really make sense in general.